### ISSUE TITLE
Prevent duplicate sibling contact capture

### ISSUE THREAD

**Issue by ChinHairSaintClair** [medic/cht-core#9601]
**Is your feature request related to a problem? Please describe.**
Currently, there appears to be no built-in deterrent against creating records with names similar to existing siblings.

**Describe the solution you'd like**
Prevent duplicate place/person creation and display possible duplicates for consideration. On record submission (through create or edit flow), we want to show the possible duplicate items to our user. They can then navigate to the possible duplicate item via a link, and proceed with record changes there, or circumvent the duplicate check & proceed with record submission.

**Describe alternatives you've considered**
Despite improving our search functionality, and training the CHWs to use it, usage of the search feature before creating new records remains low, leading to frequent duplicates.

**Additional context**
We noticed that our CHWs either forget they've previously captured items or miss previously captured items due to it being slightly mistyped. This has resulted in quite a few duplicate records being created on all user created levels of our hierarchy. We want to fix this at the source before tasks are rolled out to make sure no unnecessary/incorrect tasks fill up our CHWs worklist. This will naturally also improve the accuracy of our data for reporting purposes.

We have a working prototype that we will soon upstream, which can produce the following:
![Image](https://github.com/user-attachments/assets/fe5df62f-f56b-4dd9-89a1-268006a92824)

Please see the following related discussions for more info:
https://forum.communityhealthtoolkit.org/t/mitigate-duplicate-data-capture/3313
https://github.com/medic/cht-core/issues/6363

As a "damage control" step, as discussed with the medic team, we plan to use Databricks (our tool primarily responsible for pulling couchDB data into our community database0 to also push "flags" to potential duplicate items. That will, in turn, cause tasks to trigger on the app. CHWs are expected to then confirm/deny possible duplicate and determine what should happen to the record (delete, merge, other).

**Comment by jkuester** [comment]
## Overview

First I just want to clarify the space and existing problem-set to clarify exactly what should be addressed _in this issue_.  I think there are three separate, but related, problems to solve under the heading of "Duplicate prevention": 

1. Prevent user from creating a duplicate contact - (this current issue)
	- Runs in the webapp. Works offline.
	- Only checks against contact's *siblings* due to performance and data-access considerations.
2. Flag/merge duplicate contacts on the server - https://github.com/medic/cht-core/issues/6363
	- Sentinel transition? Historically, not possible to do much interesting here because of the limitations of Couch views.  However, [Lucene](https://github.com/medic/cht-core/issues/9542) may change the game.
	- Checks contacts across entire instance. Flags for resolution by admin (or auto-merges if possible).
3. Prevent user from creating duplicate reports - https://github.com/medic/cht-core/issues/6309
	- Proposed approach is to warn/block users when they try to create a new report for a contact that already has the same report existing in a particular time frame.

## Details

With that summary in mind, we can dig into the details of how to specifically prevent a (potentially offline) user from creating a duplicate contact.  The [prototype PR](https://github.com/medic/cht-core/pull/9609) provides a great starting point for this conversation.  My goal here is to synthesis/generalize the details from that prototype into a design summary here that is easier for folks to understand and discuss (also including some of my own suggestions and editorializing). Once we coalesce on a particular design approach, we can return to the actual code and make that happen.  

### Configuration

Since different types of contacts contain different levels of data that might need to be checked for duplication, it makes sense to configure the rules for dup checking individually for each contact type. I think it would be logical to include this config in the app-settings [`contact_types` config](https://docs.communityhealthtoolkit.org/building/reference/app-settings/hierarchy/).  

For each contact type, we need to define the rules for what constitutes a duplicate.  The most flexible way to do this would be to allow a custom function to be included in the config that accepts two contacts and returns a boolean indicating if the contacts are duplicate or not.  The Levenshtein library could be in-context for this function's logic to make use of. This would allow for the dupe logic to be as complex or simple as necessary.  The main downside of this function-based approach is that I guess it would be almost impossible to re-use this configuration for any kind of server-side dupe checking (in any future solution for https://github.com/medic/cht-core/issues/6363).  It is not feasible from a performance perspective to run a function like this _against every contact in the db_.  (Honestly, though, the more I consider this, the less I think we should try to optimize this config for any kind of server-side reuse... It seems like the most likely possibilities for server-side dupe checking functionality in the near future are not going to happen with Couch data, but via an external data store, e.g. DOT or Databricks)

The prototype PR presents an alternative approach where, instead of a function, for each contact type we define which fields should be compared from the contact docs and which algorithm (e.g. `Levenshtein` vs `NormalizedLevenshtein`) should be used to compare them.  (Note that I think it would be best to hold off on including the `queryParams` functionality or any other kind of override config in the initial MVP for duplicate checking. The goal is to add the minimum viable functionality first and then we can extend it further in future PRs.) Some considerations around this declarative approach are that it does not really allow for inter-field dependencies (e.g. specifying that contacts could have the same `name` or the same `phone` but not both).  Also, unlike the approach in the PR, we probably need to be able to define the comparison algorithm _for each field_.  Levenshtien is great for string fields, but we might need to compare things like dates or numbers in the future. 

I think more discussion about the best form of configuration would be valuable here!  One thing I would love to see is a simple way to enable some "default" dupe checks (e.g. something that just validates the `name` field).

### Functionality

Moving on to the functionality of how this should actually work in webapp.  I think the example from the PR is a solid approach where, when opening a new contact form, it also triggers a lookup of the _existing sibling contacts_ of the new contact (via `medic-client/contacts_by_parent`).  When the contact form is submitted, the configured duplicate checking is run against those sibling contacts to determine if the new contact is a duplicate of any existing ones.  Checking against the direct siblings seems like the most reasonable balance of functionality and performance.  An offline user can only know about what is visible on their device, anyway, so it is impossible to guarantee the new contact will be unique _for the whole instance_.  This means we have to stop checking somewhere.  Also, we need to do the check at the end of the form (once the new contact info has been entered), but before the form has been closed (so the user can go back and change data if necessary).  

I am not sure if this is covered in the current PR functionality or not, but I think it will be important to do the same duplicate checking _when editing a contact_.

### UX

As demonstrated in the screenshot above, the current UX in the PR is to present the user with a list of the found duplicates (along with links to go to their profile page). One incremental enhancement that might make sense here would be to present the duplicates more as a proper list of contacts (or even "contact cards") with various important identifying information displayed (instead of highlighting specifically which data is being matched for the duplicate check).  If a user clicks into one of the other contacts, they will loose all the data they have entered into the contact form, so we want to give them as much info as is feasible about the contacts before they navigate away from the form.

Instead of including the list of contacts inline in the form page, it might be better to pop a modal containing the list. 🤔  (Either way, we should be able to use a `xforms-value-changed` listener to clear the dupe error and list of contacts when the user updates a value in form.)

Another consideration is what the default behavior should be if duplicate contacts are found. Should we warn the user, but still let them submit the new contact?  Or, should we totally disable the submit button and prevent the new contact from being added?  Ultimately, this is something that we could make configurable, but it would be good to have a simple approach in the MVP and add configuration later...

---

@ChinHairSaintClair @fardarter Please weigh in here with anything that I have missed or mistaken or additional thoughts of considerations that you have!

@garethbowen let me know what you think about this proposed approach.  What other stake-holders should we pull into this conversation to make sure we can maintain momentum on this feature?

**Comment by jkuester** [comment]
> I'm really interested in seeing if we can make this workflow generic enough that it works for all report types, not just contact creation. ... For reports, can we just check report code, reported date, and subject? If we can simplify it enough, can it work out-of-the-box without configuration?

This was my initial thoughts as well, but then I was convinced by [your comment on the other issue](https://github.com/medic/cht-core/issues/6309#issuecomment-600851400) that the "duplicate report" workflow was quite different from duplicate contacts and perhaps there is not much overlap in the configs/logic.  Specifically, when detecting a "duplicate report" is is probably much less about the _contents_ of the report than just as you said: the type of report, who it is for, and when it is submitted. Basically for reports we would be looking for other reports of the same type that were created for the same contact in a particular timeframe.  These checks can happen up-front before even loading the form.  All of this is pretty different from the "duplicate contact" flow where the most important thing is the _content_ that the user enters for the new contact.  So we cannot do an upfront check for a duplicate contact. Also, it is likely that more config is necessary to allow the contact dupe checking to be really useful.  (Maybe we can find some sensible pre-sets, but is seems like lots of tuning may be needed for some cases...)

Because of this, I am skeptical of a "one-size-fits-all" solution for dupe-doc checking that covers both reports and contacts.  (And even if we do decide to go that route, we would not need to support dupe-checking reports _in this MVP PR_.)  It seems like the most important thing to decide at this point is if we think report dupe checking will need to allow for the same level of flexible configuration as contacts (e.g. specifying which fields should be dupe checked).  If so, then yeah it probably makes sense to at least design the contact dupe-checking to be extended later for also checking reports.  If not, then I think we probably just leave the report dupe checking to its own issue and not worry about it here.

> The ideal solution would warn about duplicates as early as possible. .... but implementing it as an enketo validation would be difficult I think?

Okay, this got me thinking that maybe I have been coming at this from the wrong direction!  What if, instead of configuring the dupe-checking in the app-settings, we did it in the actual form `xlsx` files?  In the form config we could use a custom column to mark all the fields that should be dupe-checked (and maybe even indicate what comparison algorithm to use).  Then we could have a custom Enekto widget that would listen for changes to any of these fields and trigger the dupe-checking logic when any of the values change. (Then, the widget _could_ trip the enketo validation logic to make the error look like a constraint violation if we wanted to go in that direction.) The data/logic flow is going to be quite a bit more complex (e.g. how to get the docs to check against, how the dupe-checking logic knows all the fields that are supposed to be included, what to do if we actually find a duplicate, etc).

The main downside I see to configuring things in the form is that for contact forms it would be important that only the fields that map directly to the contact doc are eligible for dupe checking. Fields in the `inputs` group or the `intro` group cannot be dupe-checked.  But, it seems feasible to include a validation in cht-conf to help prevent this.

(Tell me if I am wrong here, though! 😅  I feel like I am seeing Enekto widget + custom xlsx column as the solution to all problems lately....)

> I think the eCHIS Kenya team would be interested in this too.

@eljhkrr just putting this on your radar! Please jump in to the discussion here if you have any specific concerns, requirements, or ideas!

**Comment by ChinHairSaintClair** [comment]
First, thank you, @jkuester , for explaining the concept and the prototype PR so clearly.

- [Our operational needs](#our-operational-needs)
- [Critical alignment issue](#critical-alignment-issue)
- [Proposal](#proposal)
  - [Config data structure for duplicate checks](#config-data-structure-for-duplicate-checks)
  - [Duplicate conditions](#duplicate-conditions)
  - [Doing logic](#doing-logic)
  - [Marking canonical entries](#marking-canonical-entries)
  - [Defaults checks](#defaults-checks)
    - [Contact](#contact)
    - [Report](#report)
  - [Strategies](#strategies)
  - [Locations](#locations)
    - [Config](#config)
    - [Code](#code)
  - [Early duplicate detection](#early-duplicate-detection)
  - [UX](#ux)
    - [Default behaviour on duplicate found](#default-behaviour-on-duplicate-found)
    - [Display duplicates](#display-duplicates)
      - [Duplicate items as contact cards](#duplicate-items-as-contact-cards)
    - [Modification to ListTiles](#modification-to-listtiles)
- [Endnotes](#endnotes)
    - [Duplicate check on edit](#duplicate-check-on-edit)
  - [Re-fuzzy matching](#re-fuzzy-matching)


# Our operational needs

Let's headline the things we need from this solution:

1. We need the text duplicate check to handle various common sources of duplication, eg:
    a. shortened names (Kat versus Katherine, Masi versus Masiphumelele)
    b. spelling errors and casing differences.
    c. multiple valid references, eg str, str., st, st., street, ave, ave., avenue or even "ave (avenue)" or "rd (road)"
3. We need a way to mark items as duplicates/canonical.
4. We need to offer CHWs the ability to submit despite a notification that there are duplicates -- they are the expert on the ground in the moment.

# Critical alignment issue

We'll cover a number of issue in this response (thank you very much for your engagement), but what I think we're looking for with the greatest focus is rapid alignment on a [proposed data structure](#config-data-structure-for-duplicate-checks). The idea is that if we agree on a data structure, we can safely implement our own solution that will align. At worst, we might need to adjust the config load source. What we won't (we hope) need to do is a fundamental rethink/rewrite. 

This will let us progress operationally on our side (where the need is urgent) while still making space for necessary discussions around the best UX/UI, code implementation (none of this is implemented yet) and other architectural concerns. 

Please let us know if you agree with this approach or have an alternative theory.

# Proposal

## Config data structure for duplicate checks

Below is our proposed data structure for a duplicate check system. Leaving aside the question of where the config should live, a declarative data structure should be portable to `contact_types`, `<form>.properties`, or to the `.xlsx` context. There are a few elements, which we'll delve into further below, but first let me provide a broad overview.

Assume the following person-edit form:

| type              | name                      |
| ----------------- | ------------------------- |
| begin group       | person                    |
| string            | name                      |
| select_one yes_no | is_user_flagged_duplicate |
| end group         | person                    |


And its accompanying duplicate check structure:

```
{
   "props": { // The returned array will contain items that evaluate to true for the conditions - duplicates.
      "logic": "or",
      "conditions": [
         {
             "formPropPath": "/data/person/name", // The form question path whose resolved value will be compared to the corresponding value in sibling database documents.
             "dbDocRef": "name", // The name of the property to be compared in the sibling database document.
             // Strategy ALWAYS returns bool - implies status of "yes matches" or "no it doesn't match".
             "strategy": { 
                "type": "Levenshtein", // Various
                "threshold": 1.0 // Config overrides
             },
        },
        {
            "formPropPath": "/data/person/phone_number",
            "dbDocRef": "phone_number",
            "strategy": {
                "type": "Equals"
            },
        },
      ]
   },
   "shouldCheckForDuplicates": { // Optional. Defaults to "true"
      // Takes the same structure as "props" but for "dbDocRef"
      "logic": "or", 
      "conditions": [
         {
            "formPropPath": "/data/person/is_user_flagged_duplicate",
            "strategy": {
               "type": "Equal",
               "value": "no"
            }
         }
      ]
   }
}
```

## Duplicate conditions

Every condition we define is used to potentially identify a sibling database document as a duplicate. For example, if we match on the `name` property, every doc of the same type within the parent place will be flagged as a duplicate if the names are similar. These conditions can be grouped to [evaluate together](#doing-logic).

```
{
    "logic": "or",
    "conditions": [
        {
            "formPropPath": "/data/person/phone_number",
            "dbDocRef": "phone_number",
            "strategy": {
                "type": "Equals"
            }
        },
        {
            "formPropPath": "/data/person/name",
            "dbDocRef": "name",
            "strategy": {
                "type": "Levenshtein",
                "threshold": 1.0
            }
        }
    ]
}
```

|prop           | purpose                                                                                                  |
| ------------- | --------------------------------------------------------------------------------------------- |
| formPropPath  | The path of the field/question in the currently open form whose value is of interest.         |
| dbDocRef      | The property name of a sibling database document whose value is of interest.                  |
| strategy      | ALWAYS return a bool as it is a status check -- "yes, it matches" or "no, it does not match". |

The values from `formPropPath` and the sibling `dbDocRef` definitions are compared using the specified `strategy`.

> **Note**: Notice the "Equal" strategy doesn't take a value here like it does in the above `shouldCheckForDuplicates` object. This is because the current value is being compared to sibling values.

Any properties (eg `threshold`) could be provided via config, with sensible defaults.


## Doing logic

Consider the following:
| prop                     | purpose                                                                                 |
| ------------------------ | --------------------------------------------------------------------------------------- |
| props                    | Declares a set of conditions for what sibling values count as duplicates for a returned list of candidates. |
| logic                    | States how to evaluate a nested group. Support for `or`s and/or `and`s proposed. Defaults to `undefined`, but will error if `undefined` with more than one item.      |
| conditions               | Accepts groups of properties a single "duplicate condition" (as shown above), or a combination of the two for evaluation. Single items in the array will just evaluate to their value. | 
| shouldCheckForDuplicates | Prevents the duplicate check (`props`) from firing if the provided set of conditions evaluates to false. Default value is `true`. |

With the below we can, as per @jkuester 's example, identify duplicates when both the `name` and `phone_number` match a sibling's:

```
"props": {
    "logic": "or",
    "conditions": [
        {
            "logic": "and",
            // Where first_name + last_name = name
            "conditions": [
                {
                    "formPropPath": "/data/person/first_name",
                    "dbDocRef": "first_name",
                    "strategy": {
                        "type": "Levenshtein"
                    }
                },
                {
                    "formPropPath": "/data/person/last_name",
                    "dbDocRef": "last_name",
                    "strategy": {
                        "type": "Levenshtein"
                    }
                },
                {
                    "formPropPath": "/data/person/phone_number",
                    "dbDocRef": "phone_number",
                    "strategy": {
                        "type": "Equals"
                    }
                }
            ]
        },
        {
            "logic": "and",
            "conditions": [
                {
                    "formPropPath": "/data/person/first_name",
                    "dbDocRef": "first_name",
                    "strategy": {
                        "type": "Levenshtein"
                    }
                },
                {
                    "formPropPath": "/data/person/last_name",
                    "dbDocRef": "last_name",
                    "strategy": {
                        "type": "Levenshtein"
                    }
                },
                {
                    "formPropPath": "/data/person/id_number",
                    "dbDocRef": "id_number",
                    "strategy": {
                        "type": "Equals"
                    }
                }
            ]
        },
    ]
},
"shouldCheckForDuplicates": { // Optional -- defaults to "true"
    "logic": "or",
    "conditions": [
        ...
    ]
}

```

For the above example, let's imagine we have a mom and a child registered on our system. The mother, Jade, will be saved with her phone number and the child, Savanah, will share their parent's phone number. With the above config, should we attempt to create an entry of either person where **both** the `name` and `phone_number` are the same as an existing one, the duplicate list should be populated with their entry.

Notice we can also simultaneously express that we ALSO want duplicates listed where the `name` and `id_number` are the same.


## Marking canonical entries

Unfortunately, since the go-live last November, our users have created duplicates. At times, this was for legitimate reasons, such as when no house number or land parcel number (locally, ERF number) is available. The CHWs often use 'Sa number not known' in this circumstance (and type it variably). Because of elements like these, or due to typos, it is difficult to accurately search for existing documents before creating a new ones (search is also not especially accurate). 

We needed some way of identifying the gold/canonical document and marking it as canonical, so the UI should not alert about other duplicate siblings. By marking certain documents as canonical, we will be able to make decisions about updating/deleting/merging duplicates on the backend. By evaluating the current form values in the `shouldCheckForDuplicates`, we can opt out of the siblings database lookup when a document is marked as canonical.

We can express this in the same structure as in the `props` object. In this case we need to provide a comparator `value` for the `Equal` strategy (the `value`) is required, and `threshold` is overridden for `NormalizedLevenshtein` (a default could be set).

```
{
   "props":{
      "logic": "or",
      "conditions": [
         ...
      ]
   },
   "shouldCheckForDuplicates": {
      "logic": "or",
      "conditions": [
         {
            "formPropPath": "/data/person/is_canonical",
            "strategy": {
               "type": "Equal",
               "value": "no"
            }
         },
         {
            "formPropPath": "/data/person/name",
            "strategy": {
               "type": "NormalizedLevenshtein",
               "value": "Sibo",
               "threshold": 0.5 
            }
         }
      ]
   }
}
```

In the above example, we're saying that if the current form's `is_canonical` field is "yes" or if the person's name is some threshold distance from `Sibo` (just an example of the logic), the document is a known duplicate and we should not fire the database query or prompt a user to consider duplicates.

We haven't thought through all cases, but this structure may well be able to handle the inter-field deps @jkuester is thinking about.

## Defaults checks

Possible default strategies per field.

### Contact

Below is a proposal of the default strategies per field, using the `person` as an example.

```
"props": {
    "logic": "and",
    "conditions": [
        {
            "formPropPath": "/data/person/first_name",
            "dbDocRef": "first_name",
            "strategy": {
                "type": "Levenshtein",
                "threshold": 3
            }
        },
        {
            "formPropPath": "/data/person/middle_name",
            "dbDocRef": "middle_name",
            "strategy": {
                "type": "Levenshtein",
                "threshold": 3
            }
        },
        {
            "formPropPath": "/data/person/last_name",
            "dbDocRef": "last_name",
            "strategy": {
                "type": "Levenshtein",
                "threshold": 3
            }
        },
        {
            "logic": "or",
            "conditions": [
                {
                    "formPropPath": "/data/person/date_of_birth",
                    "dbDocRef": "date_of_birth",
                    "strategy": {
                        "type": "Equals",
                    }
                },
                {
                    "formPropPath": "/data/person/phone_number",
                    "dbDocRef": "phone_number",
                    "strategy": {
                        "type": "Equals",
                    }
                },
            ]
        }
    ]
}
```

### Report

Thinking about @garethbowen 's suggestions around default checks, with this language/system, consider the following default for `record_type`:

```
"props": {
    "logic": "or",
    "conditions": [
        {
            "formPropPath": "/data/report_code",
            "dbDocRef": "report_code",
            "strategy": {
                "type": "Equals",
            }
        },
        {
            "formPropPath": "/data/reported_date",
            "dbDocRef": "reported_date",
            "strategy": {
                "type": "Equals",
            }
        },
        {
            "formPropPath": "/data/subject",
            "dbDocRef": "subject",
            "strategy": {
                "type": "Equals",
            }
        }
    ]
}
```

## Strategies

Suggested strategies:

| type                    | configuration variable | example                       | reasoning                 |
| ----------------------- | ---------------------- | ----------------------------- | ------------------------- |
| `Levenshtein`           | `threshold`            | 'cat' ~ 'caterpillar' = 8     | [Fuzzy matching](https://en.wikipedia.org/wiki/Approximate_string_matching) |
| `NormalizedLevenshtein` | `threshold`            | 'cat' ~ 'caterpillar' = 0.727 | Fuzzy matching with length normalisation |
| `Equals`                | `value`                | '1994-08-23' === '1994-08-23' | Equality check. |

In a form, suppose there's a check for something like an `id_number`. This could be checked with the equality strategy.

For something like dwelling names, eg. "123 Stevenson way", "Stven Way 23", and "Peter Stevenson Road", `Levenshtein`, or `NormalizedLevenshtein` could be helpful in catching duplicates.

## Locations

Proposed storage locations for the various pieces of code that comprise this solution.

### Config

Our first suggestions for where to store config is in `app_settings.json` or in `<form>.properties.json`. We already manage form fields (listing items to hide from reports) in `<form>.properties.json` so a "duplicate" block for fields feels like a natural a fit. Any advice here would be appreciated.

Because the structure is declarative, it's possible to put the structure in the `.xlsx` files, but because it's not source controlled and adds complexity we'd resist that as an initial implementation. However, a declarative approach doesn't shut that door.

That being said, we’re a bit uncertain how to load these values and make it available within `contacts-edit.component`/`report.component`. We would appreciate any guidance to accelerate us here.

### Code

Our initial proposal is to locate all the code processing the declarations in a `modules/utils/deduplicate.ts` file, which could then be made available to `contacts-edit.component` and `reports.component`. Once again, advice here welcomed.

## Early duplicate detection

We're not exactly sure how to implement this. Maybe, since the sibling request gets triggered during form initialization, we could filter the results through `onBlur` input field listeners once it's resolved. After a certain number/set of fields are marked as dirty, we could take action. We've suggested how the results might be shown to the user, please see the [display duplicate modal](#display-duplicate-modal) point in the UX section for details.

## UX

### Default behaviour on duplicate found

It is a hard requirement for us that users can bypass the duplicate check. There may be a case for a harder check on something like reports, but that can be handled in validation, and perhaps not a default.

One idea for the UI is to add a section to the `enketo.component.html` just above the submit button with a checkbox to warn the users that there are potential duplicates, making them acknowledge the warning before moving forward. 

Here's how the flow might look:

1) On form load the sibling lookup is initiated and the submit button is disabled.
2) Users enter details that matches sibling data.
3) An inline-list is displayed, above/next to the checkbox, showing duplicate results as contact cards with detailed information and a link to navigate to the duplicate (possibly with a data loss warning).
4) A checkbox with an acknowledgement box is displayed above the submit button, which is disabled.
5) Once the checkbox is ticked, the submit button becomes enabled, allowing the form to be saved.

The benefits are:

- Flexibility for many use cases. The CHT is used by many people with different needs. 
- Minimise disruptions. Disabling the form submission could frustrate users, a warning could allow them to make an informed decision.
- Provide the field worker with the necessary agency.

### Display duplicates

> Instead of including the list of contacts inline in the form page, it might be better to pop a modal containing the list.

![Image](https://github.com/user-attachments/assets/b17278bf-141e-49e5-b681-8163e618eb4c)

An inline list of contacts (see above screenshot), allow us to leaf through the current form values and compare against the duplicates. 

Modals are great for decisions that can be made quickly where all context can be presented on the modal. However, it's clear from our users that they need to be able to compare against what is already on the page when taking decisions, and a modal flow isn't appropriate for that, especially (but not only) in multi-page registration forms.

With the current modal implementations, CHWs would have to dismiss it every time a different form value needs checking, since it blocks user interaction (no scrolling or clicking through). 

One of our suggestions for the future is to have some sort of "bottom drawer" that allows user interaction with the current form while also being able to pull up, dismiss and resize the duplicate result pane. The best of both worlds. Something like:

![Image](https://github.com/user-attachments/assets/2ab36f5b-225a-4a60-874a-e546df45d7ee)
What do you think?

#### Duplicate items as contact cards 

We like the idea of presenting duplicates as a list of "contact cards", it is something the community might be comfortable with due to familiarity. It might be advantageous to make these cards expandable (starting off with less space), since multiple duplicates could be returned. That way, as @jkuester  suggested, we could easily include various key identifiers without worrying too much about vertical space. It will allow the user to make an informed decision before navigating off the current form - losing entered data.

### Modification to ListTiles

We'd like to suggest making a subtext line available on `ListTiles`, which might be configurable through the `contact_type` section in the `app_settings`. @garethbowen  made an excellent point around assuming anything about naming conventions. We should prevent accidental duplication, not restrict intentional actions. For example, multiple family members could share the same name. 

Using the example above, adding a configurable extra line of text, like showing `date_of_birth`, could make it much easier for CHWs to tell a father and a child apart at a glance. As not all records are created during the same visit, it would be an easy way to see that the father has already been added and now it's the child's turn. It would also help them quickly spot real duplicates in their workspace.

# Endnotes

### Duplicate check on edit

> I am not sure if this is covered in the current PR functionality or not, but I think it will be important to do the same duplicate checking when editing a contact.

Short answer is yes it is. The `contacts-edit.component.ts` serves both contact creation and editing, as a result the duplicate prevention prototype was written to manage both cases. Everything works the same, just a slightly different lookup strategy on loading the page. Here's how we do it:

```
const docId = this.enketoContact.docId;
this.isEdit = docId != null;
this.parentId = contact?.parent?._id ?? this.routeSnapshot.params.parent_id;
this.contactType = contactType?.id ?? this.enketoContact?.type;
```

## Re-fuzzy matching

There are many other strategies one could consider, like the [Jaro-Winkler distance](https://en.wikipedia.org/wiki/Jaro%E2%80%93Winkler_distance) and [N-Gram Similarity](https://www.mdpi.com/2076-3417/9/9/1870), those mentioned in the strategies section are simply what're initially necessary to meet our operational requirements. We've noticed that typos are a frequent occurrence, perhaps because we have a diverse group of users with varying levels of technical know-how. A user might also use a slightly different naming scheme every other time they go out to enumerate their assigned areas. For this reason, we need a form of evaluation that allows for 'likeness', for which `Levenshtein` seems a good fit. In cases where we need to make exact comparisons, like on `id_number`, we could use `Equals`.

Forum threads that started this discussion:
https://forum.communityhealthtoolkit.org/t/unexpected-search-results/3288
https://forum.communityhealthtoolkit.org/t/mitigate-duplicate-data-capture/3313/6?u=anro

---
Special thanks to @fardarter for helping to draw this up 🥇 .

**Comment by jkuester** [comment]
Thanks for this brilliant write-up!  I have been working my way though it, making notes, and thinking about things. Unfortunately, I am OOO the rest of the week, so I do not have time for a full response here.  But, just wanted to let you know that I will be circling back here with more detailed thoughts early next week.

**Comment by ChinHairSaintClair** [comment]
The anchor tags in our responses have been fixed.

**Comment by jkuester** [comment]
> what I think we're looking for with the greatest focus is rapid alignment on a [proposed data structure](https://github.com/medic/cht-core/issues/9601#config-data-structure-for-duplicate-checks).

Agreed! Lets put the main focus on this.  I think I have two remaining functional questions here:

1. Do you foresee the need for "cross-comparing" fields as a part of the duplicate check (comparing the values of _separate_ fields). For example, comparing the `alternate_phone` value to the `phone` value of other contacts?
2. Can you confirm that the logical structuring of field comparisons you have described (e.g. a contact is duplicate if it has matching: `(first_name && last_name && phone) || (first_name && last_name && id_number)`) is a _requirement_ for your duplicate checking needs?  (As compared to a simple field list approach where a contact is duplicate if it has matching: `first_name && last_name && phone && id_number`.)

There is a ton of functionality that we _could_ build here, but I really want to try and scope things to what we actually expect to be useful...

If the answer to both `#1` and `#2` is "no", then I think our data structure could be a simple list of fields to include in the duplicate checks. (Either way, I would like to try and avoid the `formPropPath` vs `dbDocRef` duplication. The system _should_ be able to do this field mapping automatically.)

On the other hand, if we need one or both of `#1` and `#2`, then I think we should seriously consider avoiding any kind of declarative data structure for defining the duplicate checking and instead go for a Javascript expression like the existing [`context.expression` functionality](https://docs.communityhealthtoolkit.org/building/forms/app/#formsappform_namepropertiesjson) in the form `properties.json` file.  We could do a bunch of work coding up a super flexible declarative data structure and there would still be edge-cases missing and other kinds of unsupported workflows. If we need flexibility, we can have that while still keeping the code-base maintainable by supporting JS expression-style config.  To go into a bit more detail on this idea, I am thinking the duplicate_check expression could have the following things available to reference:

- `current` - the doc being created/edited
- `existing` - the existing doc to compare with (e.g. a sibling contact)
- `levenshteinEq` - function for comparing values 
- `normalizedLevenshteinEq` - function for comparing values 

Then, in your `form.properties.json` file you could have:

```json
{
    "duplicate_check": {
        "expression": "(levenshteinEq(3, current.first_name, existing.first_name) && levenshteinEq(3, current.last_name, existing.last_name)) || (current.phone === existing.phone && current.id_number === existing.id_number)"
    }
}
```

This approach is somewhat obtuse (you have to rely on the documentation to know what references are available in the expression), but I guess any approach is going to involve needing to reference docs to understand what config is available....

**Comment by jkuester** [comment]
## Defaults checks

I guess just automatically enabling some kind of default duplicate checking for all contacts could be seen as a breaking change. I wonder if it might not be better to focus on some way to allow folks to easily enable "best practice" dupe checks for each of their forms. For example, if we go with a `duplicate_check` object in the form's `properties.json` file, we could also allow other properties besides `expression`.  Even just having a `default` property on `duplicate_check` that could be set to `true` if you wanted the default duplicate check to apply to this form might work fine....

## Reports/Early duplicate detection

So far I do not perceive our design discussion as headed in a direction that would prevent us from implementing this functionality in the future.  I group these together here because I think the early duplicate detection functionality could be super useful when dupe checking reports.  (As previously noted, the most common case of "duplicate" reports would be the same `form` being submitted within a particular time window. So, we should know write at the start of a form that there are "duplicates".)

Additionally, the proposed "bottom drawer" UI could also pair well with the early duplicate detection since it would give a clean way to see the duplicate contacts while still in the middle of a form.

## Display duplicates

I really appreciate your thoughts on the modal approach and agree that an in-line/drawer UI would be superior!  My suggestion is that for the initial MVP we just display a list of the duplicate contacts in-line at the end of the form (with a few relevant details such as `name`, `phone`, `date_of_birth`, etc) with a check-box for bypassing the dupe-check. We can get fancier in future iterations.

**Comment by ChinHairSaintClair** [comment]
Thank you for your response @jkuester, we're busy working through the proposed solution and hope to have something for your consideration soon. That being said, it is the end of the year and many people are going on leave which makes it a little tricky to make meaningful progress.

# Form evolution problem

The structure and names of stored properties can vary between form versions, and this makes it difficult to guarantee field comparison between a current from and a sibling doc. We've not yet come up with a way to solve for this problem without some kind of historical mapping.

Consider the following sequence of `hhm` form versions:

First, `v1` is created with this structure:

<table>
<td style="vertical-align: top;">
    
**Create (v1)**
    
| type          | name |
| ------------- | ---- |
| `begin group` | hhm  |
| `string`      | name |
| `end group`   | hhm  |

</td>
<td style="vertical-align: top;">
    
**Edit  (v2)**

| type          | name |
| ------------- | ---- |
| `begin group` | hhm  |
| `string`      | name |
| `end group`   | hhm  |

</td>
</tr>
</table>
 
Imagine we have 10 records saved with the above structure. Then the form structure is modified (`v2`):

<table>
<td style="vertical-align: top;">
    
**Create (v2)**
    
| type            | name          |
| --------------- | ----          |
| `begin group`   | hhm           |
| `begin group`   | personal_info |
| `string`        | name          |
| `end group`     | personal_info |
| `end group`     | hhm           |

</td>
<td style="vertical-align: top;">

**Edit  (v2)**
    
| type            | name              | calculate |
| --------------- | ----              | --------- |
| `begin group`   | init              ||
| `calculate`     | prev_name         | once(coalesce(../../hhm/personal_info/name, ../../hhm/name)) |
| `string`        | name              | once(../prev_name) |
| `end group`     | init              ||
| `begin group`   | hhm               ||
| `string`        | name              ||
| `begin group`   | personal_info     ||
| `calculate`     | name              |../../init/name|
| `end group`     | personal_info     ||
| `end group`     | hhm               ||
    
</td>
</tr>
</table>

To edit a doc saved in `v1` using the `v2` edit component, should our newest reference contain no value, we fall back to the previously stored `name` from `v1` and populate the correct new field. In this example, we use the `coalesce` XPath function to check the current (`v2`) `../../hhm/personal_info/name` field for a value. If it is `null` or `""`, we fall back to the `v1` `../../hhm/name` reference. So, if `../../hhm/personal_info/name` is empty, the value from `../../hhm/name` (e.g "John Doe") is assigned to the `name` variable in the `init` group.

One advantage of this is that it gives us a progressive migration of data between form versions (the 10 records would be incrementally moved to `v2`). A significant downside is that the resolution logic gets more and more complicated the more form versions are issued with changes. The current edit form component needs to be able to resolve the correct values given incoming data saved with an arbitrary version. This is a worry for us looking forward. 

Regardless, for deduplication, the loaded siblings do not benefit from this same “migration” or mapping. 

## Approach

We’re still in the prototyping phase, but the approach we're currently playing with is to extract the property mapping from the XML and apply it to siblings before the comparison step. 

We're still not sure how complex getting those values will be.

Some expected challenges:

1) Distinguishing "mappings" from regular calculations. Let's say the field we're interested in has a calculation where a `household_name` is set to the `name` of the registered household head (should the `household_name` not be defined). No two implementers code their forms the same way, so `coalesce` might work as an identifier for us, but it might not be unique for another.
2) Loading the edit form `db doc` while in the create form flow since only the edit form contains the "mapping" info.
3) Correctly parsing the XML.

## Alternative

Consider the following structure we're looking at storing with the `expression` in the `duplicate_check` of the `.properties.json` file. A `fields` block defines the paths and db refs for the `current` and `existing` keys we're interested in for duplicate checking, without which we were left attempting to `regex` them out of the expression. We'd really like to not include paths, but the form evolution problem means we need some way of expressing the correct way to resolve the value. The only alternative we can think of at the moment is to capture the possible references as an array for each `existing` expression key.

```
{
    "expression": "levenshteinEq(3, current.first_name, existing.first_name) ..."
    "fields": {
        "current": {
            "first_name": '/data/hhm/first_name'
        },
        "existing": {
            "first_name": ['first_name', 'personal_details/first_name']
        }
    }
}
```

We don't think this is ideal as the value mapping would need to be kept in step in two different locations (.xml & .properties.json) and it adds complexity to the declaration. 

We're not sure of the best approach and wanted to expose the problem early. At some stage some new code will follow.

We would appreciate your input on this and if you have ideas for any of the other challenges.

---

# Other items

## logic

Yes, we need the ability to do logic. Let's use the `household_member` as an example, a duplicate could be identified by checking `(current.name === existing.name && current.id === existing.id) || (current.name === existing.name && current.pmi ==== existing.pmi).` 
Here, `id` is similar to a social security number, and `pmi` (patient master index) is used to link to other clinical data downstream.

## is_canonical

It's our view that `is_canonical` would make more sense as a field name to the downstream db users versus `skip_duplicate_check` as it explains the status of the doc, rather than app behaviour.

## Default checks 

How about simply setting the `"duplicate_check":true` as the default for the simple cases like checking names? The default doc for contacts could just check the `name` field, and reports could check the stricter cases that are relevant. We've not come up with a user profile that we think wouldn't want it, but are happy to defer if you know better. The only real wrench in the gears for that is the form evolution issue.

**Comment by jkuester** [comment]
> The structure and names of stored properties can vary between form versions, and this makes it difficult to guarantee field comparison between a current from and a sibling doc.

Very good point!  This pretty much solidifies it in my mind that we should avoid any declarative configuration for which fields to include in the duplicate check and should instead focus on supporting an `expression` that can be evaluated against each contact.  This will give folks the flexibility to check `current.hhm?.personal_info?.name || current.hhm?.name` as needed.  

> the approach we're currently playing with is to extract the property mapping from the XML and apply it to siblings before the comparison step.

I am still not sure we need to involve the XML/form-structure in this at all (at least not for the MVP). If we do the dupe checking down in the [FormService.saveContact method](https://github.com/medic/cht-core/blob/1033e48898305df95e151e60484d0bd1ba989e69/webapp/src/ts/services/form.service.ts#L336) after the `ContactSaveService.save` call (and maybe after applying the client-side transitions 🤔) then our `current` value in the expression can be the actual contact doc being written to the db (which should match the structure of the `existing` value).  The dupe checking logic in the `expression` will need to align with the data model of the contact _as stored in the database_.  It seems like this will greatly simplify things here. In addition to not having to try and parse the form structure, etc, it should make it trivial to handle the cases where [multiple contacts are written](https://docs.communityhealthtoolkit.org/building/forms/contact/#creating-person-and-place-contacts-in-the-same-form) in the same form.  The same goes for your alternative proposal of having the `duplicate_check.fields` values.  Ideally we avoid all that extra complexity by just allowing for the `expression` logic to operate on the actual doc data.  

The main downside of doing the dupe-checking at the doc level is that it will make it hard to know exactly which question(s) _in the form_ are involved in the duplicate violation (since that would involve the same kind of form xml parsing that I am trying to avoid).  IMHO this is okay. I don't think our MVP UX is going to try to do anything fancy here. We can show a list of the dupe contacts that were found without knowing which questions from the form were involved in the dupe check.  If we eventually do want to support some more advanced UX here, we should not be any worse off by dupe checking the docs than  we would be otherwise...

## Default behavior

After much conversation with various colleagues, my thoughts around having behavior that encourages best practices be _on by default_ has evolved.  I now believe that if we have some functionality that we believe will produce better outcomes for the majority of users (e.g. some basic dupe checking logic) we should enable that by default (and also provide a way of configuring/disabling the behavior).  In this case, we know that basic dupe-checks when creating contacts will be useful. The functionality is not going to block the user from writing contacts, so this behavior will not break any workflows one way or the other.  

What do you think about having the dupe-checking process behave like this when writing a contact?:

- If the form has `duplicate_check.expression`, use that
- If the form has nothing configured for `duplicate_check`, then we run the default dupe-checking logic
- If the form has `duplicate_check.disabled = true`, then we do not run any duplicate checking for that form.

## is_canonical

> `is_canonical` would make more sense as a field name to the downstream db users versus `skip_duplicate_check` as it explains the status of the doc, rather than app behaviour

This make total sense to me!  👍  ~~My brain is still stuck on the work "canonical", though. It is tricky for a few reasons:~~

- ~~It is not explicitly about "duplicate" checks, but could be understood to apply to other contexts.~~
- ~~When something is "canonical", it could mean that others are "non-canonical" (seems like it is often understood that only one thing at a time can be canonical). However, in this case it is likely that if we had two contacts which failed the dupe-check (e.g. a father and a son both named "Will Smith") we might end up with _both_ marked as "canonical" since we do not want to reject either as duplicate.~~

~~The problem is that I have not really been able to come up with anything better than `is_canonical` in terms of naming.  Something like `accepted/allowed_duplicate` would be more explicitly linked to the dupe checking, but I am not happy with it.  If no one can think of anything better, I am fine with `is_canonical`.~~

Edit: the longer I thought about this, the more `is_canonical` seems best.  I started thinking about it in the sense that "This Will Smith is a _canonical person_ " (and not just a duplicate record). This being different from how I was thinking previously: "This is the canonical 'Will Smith'".  

Now my only real concern here is possible collisions with existing configurations that are already using the `is_canonical` field on their contact docs.... 🤔

**Comment by ChinHairSaintClair** [comment]
# Considerations/concerns

Thank you for your response. Overall, we're happy with simplifying the solution. Below a few lingering concerns/considerations.

## Sibling lookup optimization

The `fields` block, defining each property's path, per form version, would allow us to only query the properties used in the duplicate `expression`. Reducing query reponse size, and hopefully duration. In the proposed query we use a "temporary view", or `map/reducer`, passing in the `parentId` and `contact_type` in conjunction with `fields`, to get the filtered data:

```
"fields": {
    "first_name": [
        "first_name",
        "personal_info.first_name",
        "personal.details.first_name"
    ],
    "last_name": [...]
}

...
// Above config will produce the following:
const fields = ["first_name", "personal_info.first_name", "personal.details.first_name", ...];


dbService.get()
    .query((doc, emit) => {
    if (doc.parent?._id === parentId && doc.contact_type === contactType) {
        const filteredDoc = {};

        fields.forEach(field => {
            const keys = field.split('.');
            let current = doc;
            let result = filteredDoc;

            for (const key of keys) {
                if (current[key] !== undefined) {
                    // If it's the last key, assign the value
                    if (key === keys[keys.length - 1]) {
                        result[key] = current[key];
                    } else {
                        // Ensure intermediate object exists in result
                        result[key] = result[key] || {};
                        result = result[key]; // Move deeper into the result
                        current = current[key]; // Move deeper into the source doc
                    }
                } else {
                    // Key doesn't exist - exit the loop
                    break;
                }
            }
        });

        emit(null, filteredDoc);
    }
});
```

Without knowing the comparison properties upfront, we're left querying the entire contents of each sibling document. We're worried this may have a significant impact on low spec devices, especially those with large datasets.

Our CHWs often have 300+ documents, `Plots`, captured within their assigned `Indawo` (area), each with the following content:

```
{
  "_id": "73207713-4b08-4a58-a3d9-549e37dd34ad",
  "_rev": "1-1cdf6bd74f181d9fe7e7c2f249dcaedf",
  "parent": {
    "_id": "fa29b53a-a5a7-4548-a5a6-ae0c8492669a",
    "parent": {
      "_id": "0a69e4fc-4b99-4b0c-8e1c-a003fc197e3a",
      "parent": {
        "_id": "0c154d7c-f7ff-4d04-85a3-7f91d4cf7289"
      }
    }
  },
  "type": "contact",
  "visit_date": "2024-12-17",
  "name": "plot",
  "description": "",
  "postal_code": "7570",
  "location": "",
  "gps_issue": "",
  "gps_issue_other": "",
  "visit_outcome": "yes",
  "visit_outcome_other": "",
  "dwelling_type": "residential",
  "dwelling_type_other": "",
  "residential_type": "informal_dwelling",
  "health_service_type": "",
  "personal_services_type": "",
  "g_retail": {
    "retail_type": "",
    "shop_type": "",
    "food_stores_type": ""
  },
  "g_entertainment": {
    "entertainment_type": "",
    "social_type": "",
    "sports_type": ""
  },
  "g_industrial": {
    "industrial_type": "",
    "workshop_type": ""
  },
  "office_type": "",
  "education_type": "",
  "g_travel_related": {
    "travel_related_type": "",
    "transport_type": "",
    "short_stay_residential_type": ""
  },
  "g_public_services": {
    "public_services_type": "",
    "other_government_services_type": "",
    "parastatal_type": "",
    "community_services_type": ""
  },
  "outbuilding_type": "",
  "inaccessible_type": "",
  "open_space_type": "",
  "has_second_use": "no",
  "second_use": {
    "dwelling_type": "",
    "dwelling_type_other": "",
    "residential_type": "",
    "health_service_type": "",
    "personal_services_type": "",
    "g_retail": {
      "retail_type": "",
      "shop_type": "",
      "food_stores_type": ""
    },
    "g_entertainment": {
      "entertainment_type": "",
      "social_type": "",
      "sports_type": ""
    },
    "g_industrial": {
      "industrial_type": "",
      "workshop_type": ""
    },
    "office_type": "",
    "education_type": "",
    "g_travel_related": {
      "travel_related_type": "",
      "transport_type": "",
      "short_stay_residential_type": ""
    },
    "g_public_services": {
      "public_services_type": "",
      "other_government_services_type": "",
      "parastatal_type": "",
      "community_services_type": ""
    },
    "outbuilding_type": "",
    "inaccessible_type": "",
    "open_space_type": ""
  },
  "has_third_use": "",
  "third_use": {
    "dwelling_type": "",
    "dwelling_type_other": "",
    "residential_type": "",
    "health_service_type": "",
    "personal_services_type": "",
    "g_retail": {
      "retail_type": "",
      "shop_type": "",
      "food_stores_type": ""
    },
    "g_entertainment": {
      "entertainment_type": "",
      "social_type": "",
      "sports_type": ""
    },
    "g_industrial": {
      "industrial_type": "",
      "workshop_type": ""
    },
    "office_type": "",
    "education_type": "",
    "g_travel_related": {
      "travel_related_type": "",
      "transport_type": "",
      "short_stay_residential_type": ""
    },
    "g_public_services": {
      "public_services_type": "",
      "other_government_services_type": "",
      "parastatal_type": "",
      "community_services_type": ""
    },
    "outbuilding_type": "",
    "inaccessible_type": "",
    "open_space_type": ""
  },
  "has_fourth_use": "",
  "fourth_use": {
    "dwelling_type": "",
    "dwelling_type_other": "",
    "residential_type": "",
    "health_service_type": "",
    "personal_services_type": "",
    "g_retail": {
      "retail_type": "",
      "shop_type": "",
      "food_stores_type": ""
    },
    "g_entertainment": {
      "entertainment_type": "",
      "social_type": "",
      "sports_type": ""
    },
    "g_industrial": {
      "industrial_type": "",
      "workshop_type": ""
    },
    "office_type": "",
    "education_type": "",
    "g_travel_related": {
      "travel_related_type": "",
      "transport_type": "",
      "short_stay_residential_type": ""
    },
    "g_public_services": {
      "public_services_type": "",
      "other_government_services_type": "",
      "parastatal_type": "",
      "community_services_type": ""
    },
    "outbuilding_type": "",
    "inaccessible_type": "",
    "open_space_type": ""
  },
  "expected_household_count": "1",
  "notes": "",
  "geolocation": "",
  "meta": {
    "created_by": "admin",
    "created_by_person_uuid": "",
    "created_by_place_uuid": ""
  },
  "form_start": "2024-12-17T12:06:37.975+02:00",
  "contact_type": "dwelling",
  "reported_date": 1734430010327,
  "form_version": {
    "time": 1734429915179,
    "sha256": "8019cd7b229514bb71de1790e594612ed3eba0008afb514366165105420df806"
  }
}
```

In addition to this, the entire object is passed into the parser when evaluating the expression. For every sibling.

Regex-ing the fields from the expression is not a viable option in our view. If we're to do this, we need to have the necessary fields/properties declared - as seen in the example above).

It is important to note that the query in the above example will only work for `offline` users since `pouchDB` allows "temporary views", but not `couchDB`.

We could circumvent this issue by using the existing session service that is injected at root:

```
import { SessionService } from '@mm-services/session.service';
...
constructor(
    private sessionService:SessionService,
...
this.isOnlineOnly = this.sessionService.isOnlineOnly();
```

And use the persisted view query for online users since most would be on PCs with enough horsepower to not be as impacted:

```
this.dbLookupRef = this.dbService
    .get()
    .query('medic-client/contacts_by_parent', { // Existing CHT view
      startkey: [this.parentId, this.contactType],
      endkey: [this.parentId, this.contactType, {}],
      include_docs: true
    });
```

That being said. We're unsure of the performance impact of a temporary view vs a persisted view, and if that impact is negligible compared to the gains of reduced returns. We would appreciate your input on this.


## Expression complexity

A note on the suggested use of expressions: having each property's fallback value listed in the expression could make it difficult to read and maintain. Especially as forms evolve over time. Consider the following expression:

```
(levenshteinEq(current.first_name, existing.first_name || existing.personal_info.first_name) && levenshteinEq(current.last_name, existing.last_name || existing.personal_info.last_name) && current.phone_number === (existing.phone_number || existing.personal_info.phone_number)) || current.id_number === (existing.id_number || existing.personal_info.id_number)
```

versus (a possible):

```
{
    "expression": "(levenshteinEq(current.first_name, existing.first_name) && levenshteinEq(current.last_name, existing.last_name) && current.phone_number === existing.phone_number) || current.id_number === existing.id_number",
    "field_resolvers": {
        "first_name": "existing.first_name || existing.personal_info.first_name",
        "last_name": "existing.last_name || existing.personal_info.last_name",
        "phone_number": "existing.phone_number || existing.personal_info.phone_number"
        "id_number": "existing.id_number || existing.personal_info.id_number"
    }
}
```

The properties.json space won't provide any code completion or syntax hints, and that might be difficult to spot mistakes. Moving the `existing` prop paths into a separate `field_resolvers` space could help with readability and maintainability.

> Important: This still may not be viable if we need to pass values into a query.

## Duplicate item display

Without having the priority properties or fields defined, we'd be unable to decern the critical info from the fluff. Each sibling's full content will be dumped into a display item as is. 

Imagine the scenario where we returned siblings with as much properties as shown above. Wading through 3+ siblings, each containing all that info, to pinpoint the property causing the duplication warning would be terribly cumbersome. Especially since initially we'll lack the capabilities to collapse/expand. It's our view that we need to find some way of articulating what fields we're interested in. 


## Move code to FormService.saveContact

We're in favour of this. Accessing the dedup code on API level is probably better than on the UI. After the transitions fire, the doc would be in its most complete state before considering duplication and there would be no need to do custom parsing of the XML. That said, if the moving of dedup code also implies the sibling lookup, we'll lose out on the request completing in the background while the user is busy filling in the form. Since the lookup will now fire on submit. It could have a considerable impact on slower devices or those with larger datasets. 


# Multiple contacts per form

We're not sure why this should be consideration as part of the duplicate prevention. Aren't multiple contacts only created on `PLACE` create forms? Since this `PLACE` will be a new entry, and if it passes duplicate checks, all its children will be unique. The validation of the children would be handled on the form itself. Please let us know if we're misunderstanding.


# is_canonical

Agreed. We could follow a similar approach as with the xPath functions where we prefix the `cht` keyword. What are your thoughts?

# Default behaviour

We like the proposed `duplicate_check` default implementation flags/behaviour.

**Comment by jkuester** [comment]
## Sibling lookup optimization

I agree that a focus on performance is super important here. However, in my experience it can be very tricky to optimize Couch/Pouch stuff because a lot of things I assume would help improve performance actually an make things worse...  For the "temporary views", the [Pouch docs](https://pouchdb.com/guides/queries.html#mappin-and-reducin) say:

> Temporary queries are very slow, and we only recommend them for quick debugging during development. 

I think the problem is that they are not indexed and so basically trigger a full DB scan every time they are used.  In a perfect world, [mango queries](https://pouchdb.com/guides/mango-queries.html) might be useful here since we could easily create mango indexes at runtime, but usage of [mango indexes in the CHT](https://github.com/medic/cht-core/issues/5592) is not going to happen soon and I don't think we should wait on it here.  😞 

On the other hand, using a persisted view (e.g. `contacts_by_parent`) tends to perform very well even when used offline on low-speced devices. The only question is how fast could we spin through a large list of sibling docs checking for dups.  In the most extreme worst-case scenario I guess a contact could have "thousands" of siblings, but probably not "tens-of-thousands". In most cases, though, I would expect the number to never be higher than "hundreds".  At that level, I hesitate to prematurely optimize too much. If a particular partner's contact hierarchy causes terrible performance here, they could work around it by just disabling the dupe checking (and at that point we can revisit this for more optimizations).

## Expression complexity

My concern with the `field_resolvers` approach here is that in some sense we are just spreading out the complexity across multiple inter-dependent JSON fields (and the dev experience is still bad because, like you pointed out, there is no IDE support).

One alternative would be to follow the pattern we have in the config for the `contact_summary` calculation.  The code to calculate the contact summary is copied into the `contact_summary` string field in the `app_settings.json` file from the `contact-summary.templated.js` file by the cht-conf `compile-app-settings` command.  You do not have to hand-write the contact summary code directly into the JSON field.  In the same way, we could have a file that lived next to the forms (something like `district_hospital.duplicate_check.js`).  Then the cht-conf `convert-contact-forms` action could automatically populate the `duplicate_check.expression` field in the `properties.json` files with data from the `duplicate_check.js` files....

IMHO this is something we could add support for in the future, but would not necessarily need to include in the MVP.  We could start with support for simple `expression` values directly in the properties files. Then, if/when we have a need for more complex code, we can extend support for an actual Javascript file. 

## Duplicate item display

> Without having the priority properties or fields defined, we'd be unable to decern the critical info from the fluff. Each sibling's full content will be dumped into a display item as is.

Right, we defintily don't want to do this.  My thought here was that it might not be terribly useful to the user to know the _exact questions/fields_ which contributed to finding the duplicate.  Instead, the user should just see a list of contacts the system as determined to be "duplicate" (similar to what is shown in [this prototype screenshot](https://private-user-images.githubusercontent.com/41445931/390403194-2ab36f5b-225a-4a60-874a-e546df45d7ee.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzNTE3MDEsIm5iZiI6MTczNjM1MTQwMSwicGF0aCI6Ii80MTQ0NTkzMS8zOTA0MDMxOTQtMmFiMzZmNWItMjI1YS00YTYwLTg3NGEtZTU0NmRmNDVkN2VlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDE1NTAwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFhMWViYmM0OGUzYTFhMjZhZGY1OTFiMDE3NzRmOTg5ODQzZWI3ODVhZjkwOTdiZmQyNTEzMDk1ZTg0YzkzNGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GbXbBe9sOUGbgeyEnqCK5k0Wljax5ZRWFbHEJ7WNO2s)).  

To me it seems like the question the user is going to need to answer is not "which field can I change so my new contact is not considered a duplicate?" but instead is just "is my new contact actually a canonical person or are they already represented by one of these other contacts?". Does that make sense?  (Hoping to find the most simple UX as possible while still making sure the user has what they need to make a decision.)

A simple contact list (maybe with name/dob) seems adequate for an MVP, but I could see adding additional functionality in the future so that a user could tab a list entry and it would pop a model (while still remaining in the form) with additional details about the contact. This would let the user do a bit more investigation into the duplicates before leaving the form.

## Move code to FormService.saveContact

>  That said, if the moving of dedup code also implies the sibling lookup, we'll lose out on the request completing in the background while the user is busy filling in the form. Since the lookup will now fire on submit. It could have a considerable impact on slower devices or those with larger datasets.

Yeah, this is a good point. However, I expect we could find some way to supply the looked-up siblings to the dup check code even down in the FormService (technically, the `FormService.render` function is going to be called each time a form is _created_ so we could possibly even tie in something there...).

A more troublesome question, though, is if we _should_ be pre-loading the sibling data, or if we need need to wait until the form is submitted before getting the sibling data. The case I am considering is when you have an online user creating a contact. They open the form, but maybe they take awhile to complete it and hit the submit button. In the meantime, a _separate user_ creates/updates a contact in the DB such that it should be a duplicate of the original user's new contact.  When the original user finally submits their form, if we dupe check against sibling docs that we looked up when the form opened, we would miss the fact that the new contact was now actually a duplicate....

Regardless of what we do, there is no feasible way to make these dupe-checks atomic with the database _writes_.  Especially with offline users, I think it would be a mistake to expect this dupe-check functionality to be _guaranteeing_ data integrity. (Instead I see it as more of a user aid to help them make the right decisions.) It will always be possible for a scenario to happen were we still end up writing contacts that should be considered duplicates.  So, with that in mind, perhaps it is better to prioritize performance for this functionality and stick with the approach of loading the siblings when the form is created...  I can see it both ways. What makes the most sense to you?

## Multiple contacts per form

I think you are totally correct here.  I was zoomed in too close looking at the code and not considering the broader workflows here.  😬 Sorry for the added noise!  

## is_canonical

`cht_is_canonical` could work. 🤔 The more I think about this, the weirder the scenarios get to me.  The whole fully-configurable schema-less data model just makes all this sketchy.  😓   I know some languages/conventions use underscores to indicate fields used by the system/framework (e.g. Couch has `_id` and `_rev`). So `_is_canonical` is another option.  

What I am stuck on, though, is that up until now, we have not used any special conventions for this kind of thing (even though a number of fields such as `name` and `date_of_birth` are very important to the CHT logic. My proposal is to just move forwards with `is_canonical` for now.  I have created a [forum thread](https://forum.communityhealthtoolkit.org/t/how-to-differentiate-between-custom-fields-in-documents-and-fields-used-by-the-cht/4547) specific to this question and we can continue the discussion out there....

---

I feel like we are getting very close to a clear design here. For any more open questions, I am happy to continue this conversation async here as we have been doing. Alternatively, if you think it would be helpful I could organize a call where we could hash through any final discussions "in person". 

Ultimately, if you feel comfortable with the overall structure described here, maybe it makes sense to just progress to a PR?

**Comment by ChinHairSaintClair** [comment]
Thank you for taking the time to address each concern, @jkuester. We've updated the linked PR accordingly and would appreciate your feedback. For now, all duplicate-related code has been moved to the `form.service`, including the sibling lookup. If performance becomes an issue, we can always move the lookup back to the `contacts.edit.component`. We've also added basic expand/collapse functionality for the duplicate item details.

Two issues remain:
1. **Circular navigation issue:**
Navigating to a duplicate item creates a circular navigation problem in the back-stack. Clicking the "Take me there" button on a duplicate item works as expected, and pressing the `back` button takes you back to the "current" form. However, if you press `back` again, instead of returning to the "profile" page where the "+" button was clicked, you're navigated back to the duplicate item.
I expected that pressing `back` would pop the duplicate item from the back-stack. On mobile, it seems a `replace` approach is taken during navigation due to how the back button works. Perhaps we should replicate this behaviour for desktop? Instead of using `this.router.navigate(['/contacts', _id, 'edit']);`, we could try `this.router.navigate(['/contacts', _id, 'edit'], { replaceUrl: true });`. What do you think?
2. **SonarCloud flagging:**
SonarCloud flags `Object.prototype.hasOwnProperty(...)` in favour of `Object.hasOwn`, but the former is consistent with the rest of the codebase. Unfortunately, `Object.hasOwn` isn't available due to the ECMAScript version used in CHT.

**Comment by jkuester** [comment]
👍 Awesome!  I am going to be out for the next few days, but definitely plan to dig into this when I am back next week.  Thank you!
